lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE LambdaCase #-}
+
+module LambdaBuffers.Compiler.TypeClass.Solve (solve) where
+
+import Data.List (foldl', sortBy)
+import Data.Text (Text)
+import LambdaBuffers.Compiler.TypeClass.Pat (
+  Pat (AppP, DecP, ProdP, RecP, RefP, SumP, VarP, (:*), (:=)),
+  matches,
+ )
+import LambdaBuffers.Compiler.TypeClass.Rules (
+  Class (csupers),
+  Constraint (C),
+  Rule ((:<=)),
+  mapPat,
+  ruleHeadClass,
+  ruleHeadPat,
+ )
+
+import Data.Set qualified as S
+
+{- Variable substitution. Given a string that represents a variable name,
+   and a type to instantiate variables with that name to, performs the
+   instantiation
+-}
+subV :: Text -> Pat -> Pat -> Pat
+subV varNm t = \case
+  var@(VarP v) -> if v == varNm then t else var
+  x :* xs -> subV varNm t x :* subV varNm t xs
+  l := x -> subV varNm t l := subV varNm t x
+  ProdP xs -> ProdP (subV varNm t xs)
+  RecP xs -> RecP (subV varNm t xs)
+  SumP xs -> SumP (subV varNm t xs)
+  AppP t1 t2 -> AppP (subV varNm t t1) (subV varNm t t2)
+  RefP n x -> RefP (subV varNm t n) (subV varNm t x)
+  DecP a b c -> DecP (subV varNm t a) (subV varNm t b) (subV varNm t c)
+  other -> other
+
+{- Performs substitution on an entire instance (the first argument) given the
+   concrete types from a Pat (the second argument).
+   Note that ONLY PatVars which occur in the Instance *HEAD* are replaced, though they
+   are replaced in the instance superclasses as well (if they occur there).
+-}
+subst :: Rule -> Pat -> Rule
+subst cst@(C _ t :<= _) ty = mapPat (go (getSubs t ty)) cst
+  where
+    go :: [(Text, Pat)] -> Pat -> Pat
+    go subs tty =
+      let noflip p1 p2 = uncurry subV p2 p1
+       in foldl' noflip tty subs
```

Comment:
This is very much unification.

Reply:
I really don't think it is. Unification is more than just substitution. I'm not doing unification if I have a logical sentence 

```
P -> (P v Q)
```
and I substitute `R & S` into `P` yielding 

```
(R&S) -> ((R&S) v Q)
```

Reply:
You have to understand what the point of this is. In practice the ONLY point of `subst` is to substitute into type variables in the "constraints" part of a `Rule`. We do it to the head as well because... well, we might as well I guess, but it's always a unidirectional thing. 

If the head of a rule matches a `Pat` then we don't actually care about the head, but we need to make sure the pattern variables in the constraints are instantiated to the right thing

Reply:
Really this *can't* be unification.... if for no other reason than this was well established 100+ of years before unification existed. If this were unification then Herbrand didn't discover anything when he discovered unification lol

Reply:
I won't spend forever on this since it doesn't really matter much, but there's a way of understanding how this works that looks a lot more like case expressions than like unification. The `Rule`s can be thought of as patterns (in the Haskell sense, the thing that goes to the left of `->` in the body of a case expression) and the `Pat` in the head of the constraint we're solving for can be thought of as an expression. 

So if we have 

```
scope: 
rule C Int 
rule C a => C (Maybe a)
```

what that "means" is something like 

```
solve x = case x of 
  C Int -> ()                                             -- Branch A 
  C (Maybe patVar) -> solve patVar   -- Branch B 
```

So, concretely w/ `C (Maybe Int)`: 

```
solve (C  (Maybe Int))
   > solve Int                     -- matches Branch B 
   > ()                                  -- matches Branch A 
```

And substitution (which is always substituting a `C (Maybe Int)` _expression-ey_ thing into a more general pattern w/ a patVar "hole" like `C (Maybe a)`) is just how we determine whether something matches the pattern. `C Int` is effectively a literal pattern that matches one specific expression.

Reply:
Or, to try another way of putting it: It's not unification b/c the solution to a unification problem involves a uniform assignment of terms to variables _on both sides_ of the equation. But we only "assign" terms to variables on the left hand (more general) side. 

If I have a rule `C (F a)` and I'm trying to solve `C (F (G b))`, I'll "assign" `F (G b)` to `a` but I'll **NEVER** assign anything to `b`. 

We are assigning variables uniformly to both sides of the `=>` in a rule, but `=>` doesn't express an equality so even there we're just doing substitution.

Reply:
c.f. [this better explanation of the difference between pattern matching and unification](https://pages.cpsc.ucalgary.ca/~robin/class/449/notes/MatchUnif.pdf)

Variables on the "left hand side" (i.e. in Rules) are what you might call "unification variables" but there are no unification variables on the right hand side. If one side by definition (or artifice in this case) doesn't have unification variables then we're doing pattern matching.

...which is admittedly somewhat obscured by my implementation here since `VarP` on the Rule side means something different than `VarP` on the right hand side.

Reply:
And, therefore, `subst` doesn't do unification. `Rule`s are always the "template" and bare `Pat`s are always the "subject". There _aren't_ "templates" and "subjects" in Prolog the way there are in pattern matching expressions in Haskell (or other languages).

Reply:
> This is very much unification.


> As I said, just by having ground terms on one side you can call it pattern matching, I'm still saying, this is AT LEAST covered by unification.

"X is Y" and "X is covered by Y" are two very different statements. You can of course implement pattern matching w/ a unification algorithm. But you do not _have to_. Unification is a more general / powerful / difficult to implement procedure than pattern matching. My references to the Haskell internals were just to demonstrate that you can implement a perfectly satisfactory pattern matching algorithm in a simpler way.